File: OOP/08.OOP in Java/src/com/softuni/oop/geometry/Rectangle.java
package com.softuni.oop.geometry;

public class Rectangle extends PlaneShape {
    private double width;
    private double height;

    public Rectangle(Point2D vertexA, double width, double height) {
        super(vertexA);
        this.setWidth(width);
        this.setHeight(height);
    }

    public double getWidth() {
        return this.width;
    }

    public double getHeight() {
        return this.height;
    }

    private void setWidth(double width) {
        if (width <= 0.0) {
            throw new IllegalArgumentException("Invalid width - cannot be of negative or zero value.");
        }

        this.width = width;
    }

    private void setHeight(double height) {
        if (height <= 0.0) {
            throw new IllegalArgumentException("Invalid height - cannot be of negative or zero value.");
        }

        this.height = height;
    }

    @Override
    public double getArea() {
        return this.getWidth() * this.getHeight();
    }

    @Override
    public double getPerimeter() {
        return 2 * (this.getWidth() + this.getHeight());
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        output.append("Shape type: Rectangle\n");
        output.append(super.toString());

        return output.toString();
    }
}

File: OOP/06.Encapsulation and Polymorphism/01.Shapes/Rectangle.cs
namespace Shape
{
    using System;

    public class Rectangle : BasicShape
    {
        public Rectangle(double width, double height)
            : base(width, height)
        {
        }

        public new double Width
        {
            get
            {
                return base.Width;
            }

            set
            {
                base.Width = value;
            }
        }

        public new double Height
        {
            get
            {
                return base.Height;
            }

            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("value", "Side cannot be zero or negative.");
                }

                base.Height = value;
            }
        }

        public override double CalculateArea()
        {
            return this.Width * this.Height;
        }

        public override double CalculatePerimeter()
        {
            return (2 * this.Width) + (2 * this.Height);
        }
    }
}
